Refactor GitOps approach prompt flow in guided init

marcelovilla opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://nebari.dev/community
-->

## Reference Issues or PRs
Closes #2268 

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create a link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

## Any other comments?

## How to test this PR
While we have unit tests covering the functionality changes in this PR, I present three different paths a reviewer might want to test manually.

#### Choose GitLab during the GitOps approach prompts
1. Run `nebari init --guided-init`
2. Follow through with any options and select `y` when prompted with `Would you like to adopt a GitOps approach to managing Nebari? (y/N)`
3. Choose `gitlab.com`

You should be prompted with further configuration options. Follow through selecting `N` for the following questions.

4. Run `nebari render -c nebari-config.yaml`

You should see a `.gitlab-ci.yml` file.

#### Choose GitHub during the GitOps approach prompts
1. Run `nebari init --guided-init`
2. Follow through with any options and select `y` when prompted with `Would you like to adopt a GitOps approach to managing Nebari? (y/N)`
3. Choose `github.com`
4. Choose `N` when prompted with `Would you like nebari to create a remote repository on github.com?`

You should be prompted with further configuration options. Follow through selecting `N` for the following questions.

5. Run `nebari render -c nebari-config.yaml`

You should see a `.github/workflows` folder.

#### Choose GitHub with auto-provisioning during the GitOps approach prompts
1. Run `nebari init --guided-init`
2. Follow through with any options and select `y` when prompted with `Would you like to adopt a GitOps approach to managing Nebari? (y/N)`
3. Choose `github.com`
4. Choose `y` when prompted with `Would you like nebari to create a remote repository on github.com?`
5. Pass an organization you have access to
6. Pass a repo name to be created
7. Paste your github username and token if necessary

You should be prompted with further configuration options. Follow through selecting `N` for the following questions.

5. Run `nebari render -c nebari-config.yaml`

You should see a `.github/workflows` folder and there should be newly created GitHub repo under the organization you specified.

<!--
Please be aware that we are a loose team of volunteers, so patience is necessary;
assistance handling other issues is very welcome.
We value all user contributions. If we are slow to review, either the pull request needs some benchmarking, tinkering,
convincing, etc., or the reviewers are likely busy. In either case,
we ask for your understanding during the
review process.
Thanks for contributing to Nebari 🙏🏼!
-->

marcelovilla commented:
Unit tests are failing. I'm looking into it.

aktech commented on src/nebari/schema.py:
```diff
@@ -17,7 +17,7 @@
 email_pydantic = pydantic.constr(regex=email_regex)
 
-github_url_regex = "^(https://)?github.com/([^/]+)/([^/]+)/?$"
+github_url_regex = "^(https://)?(github.com|gitlab.com)/([^/]+)/([^/]+)/?$"
```
How do you feel about writing something like:

```python
git_repo_url_regex = "^(https://)?(git)(hub|lab)(.com)/([^/]+)/([^/]+)/?$"
```

- to remove repetition of common string `git` and `.com`
- also it's no more just github :)

marcelovilla replied:
Thanks, Amit! That is indeed much nicer. I just made a slight modification at the end to avoid allowing urls with a trailing slash that do not have a repo name. For example:
```
https://gitlab.com/user/
```

This was not being caught before.

aktech reviewed (changes requested):
I have little suggestion for improving the regex. I would also suggest writing some tests, like for example:

```python
import re
import pytest

@pytest.mark.parametrize("url", [
    "https://github.com/user/repository",
    "https://gitlab.com/user/repository/",
])
def test_valid_git_repo_urls(url):
    assert re.match(git_repo_url_regex, url) is not None, f"Failed to match {url}"

@pytest.mark.parametrize("url", [
    "https://github.com",
    "https://github.com/user",
    "https://gitlab.com/user",
    "http://github.com/user/",
    "gitlab.com/user/",
    "github.com",
])
def test_invalid_git_repo_urls(url):
    assert re.match(git_repo_url_regex, url) is None, f"Unexpected match for {url}"

```

marcelovilla commented:
@aktech I modified the existing tests and added new ones based on your suggestion. I should have added new tests in the first place so thanks for bringing that up!

aktech commented on tests/tests_unit/test_cli_init_repository.py:
```diff
@@ -174,21 +174,45 @@ def test_cli_init_error_repository_missing_env(monkeypatch: pytest.MonkeyPatch):
+    ],
+)
+def test_cli_init_valid_repo(url):
+    app = create_cli()
+
+    args = ["init", "local", "--project-name", "test", "--repository", url]
+
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp_file = Path(tmp).resolve() / "nebari-config.yaml"
+        assert tmp_file.exists() is False
 
+        result = runner.invoke(app, args + ["--output", tmp_file.resolve()])
+
+        assert 0 == result.exit_code
+        assert not result.exception
+        assert tmp_file.exists() is True
```
This code is very similar to the one below, how do you feel about creating a function and reusing that?

dcmcand commented:
@marcelovilla can you update the status here?

dcmcand commented:
@marcelovilla what is the status of this PR? Are you still working on it?

marcelovilla commented:
This PR was stale for a very long time but I've just added some changes to get it across the finish line. You'll see that I decided to revert the regex changes I wrote before (and thus, also decided to change the title of the PR). The reason behind this is that I did not see any point in modifying the regex to allow both GitHub and GitLab repos when we don't really support repo auto-provisioning for GitLab.

I decided to keep only the regex for GitHub repo validation and instead change the flow in which we present the user with the GitOps approach prompts during the guided init. Previously, we were prompting the user for the following inputs:

1. Provider (i.e., GitHub or GitLab)
2. Organization and repository
3. Whether they want Nebari to auto-provision the repository

When users selected GitLab for (1) and then gave an org and repo name for (2), they would get a validation error because of the repo URL (see #2268). While having a regex that validated both provider repos would have solved that issue, it wouldn't really have done anything with that information afterwards, because, as I mentioned before, we have not implemented the auto-provisioning logic for GitLab. I think implementing that is outside of the scope of this PR and will require a considerably larger effort to achieve. Furthermore, in the case where users choose GitHub as the provider, we only use the organization and repository name if they want Nebari to auto-provision the repository. Thus, with this PR my proposal is to modify the GitOps approach flow in the guided init to prompt the user for the following inputs:

1. Provider (i.e., GitHub or GitLab)
2. (Only if the provider is GitHub) whether they want Nebari to auto-provision the repository
3. (Only if they want to auto-provision the repo) organization and repository

Additionally, I modified our render tests to make sure we're passing GitLab as a provider in one of the configurations. And unrelated to this, I fixed the original regex to avoid matching an edge case, which I added to the existing tests.

The PR description has a section on how to test this PR.

Adam-D-Lewis reviewed (approved):
LGTM

marcelovilla commented:
Thanks for the review @Adam-D-Lewis

@aktech I just realized I cannot merge this as you requested changes a while ago. Can you have a look and see what you think?

marcelovilla commented:
@aktech just wanted to give this PR a nudge when you have some time. Thanks!

aktech reviewed (approved):
Looks good! 👍
